### ISSUE TITLE
cmd/trace: runs out of memory processing 30 MB file

### ISSUE THREAD

**Issue by AlexRouSg** [golang/go#21870]
Please answer these questions before submitting your issue. Thanks!

### What version of Go are you using (`go version`)?
go version go1.9 linux/amd64

### Does this issue reproduce with the latest release?
Yes

### What operating system and processor architecture are you using (`go env`)?
GOARCH="amd64"
GOBIN=""
GOEXE=""
GOHOSTARCH="amd64"
GOHOSTOS="linux"
GOOS="linux"
GOPATH="~/go"
GORACE=""
GOROOT="~/.local/go"
GOTOOLDIR="~/.local/go/pkg/tool/linux_amd64"
GCCGO="gccgo"
CC="gcc"
GOGCCFLAGS="-fPIC -m64 -pthread -fmessage-length=0 -fdebug-prefix-map=/tmp/go-build582547481=/tmp/go-build -gno-record-gcc-switches"
CXX="g++"
CGO_ENABLED="1"
CGO_CFLAGS="-g -O2"
CGO_CPPFLAGS=""
CGO_CXXFLAGS="-g -O2"
CGO_FFLAGS="-g -O2"
CGO_LDFLAGS="-g -O2"
PKG_CONFIG="pkg-config"

### What did you do?
go tool trace trace.out

### What did you expect to see?
2017/09/14 03:16:35 Parsing trace...
2017/09/14 03:16:40 Serializing trace...
2017/09/14 03:16:41 Splitting trace...
2017/09/14 03:16:47 Opening browser

### What did you see instead?
2017/09/14 04:31:17 Parsing trace...
2017/09/14 04:31:32 Serializing trace...
go tool trace: signal: killed

Something to note, OS froze (keyboard/mouse unresponsive) for a few seconds before trace terminated.

[trace.out.tar.gz](https://github.com/golang/go/files/1300898/trace.out.tar.gz)

**Comment by AlexRouSg** [comment]
So I ran it with the ram/cpu monitor opened and noticed it was eating 100% ram before my linux box froze. Added some swap and looks like it ate 6 GB of ram to open the webpage and then trying to view a trace ate another 1.5 GB - 2 GB and then froze my linux box without showing the graph.

Do these numbers sound right? They are from looking at the total ram +swap usage so it's not too accurate.

**Comment by davecheney** [comment]
Yes. Traces can be very large.

**Comment by AlexRouSg** [comment]
Any options? Or am I just screwed now cause this is from just 5 seconds of run time.

**Comment by davecheney** [comment]
How much ram does your host have?

**Comment by AlexRouSg** [comment]
8 GB, but even with swap I couldn't get the graphs to load before it freezes.

**Comment by davecheney** [comment]
/cc @francesc

TL;DR you need more ram. Sorry.

**Comment by AlexRouSg** [comment]
Was expecting that...
But if this is from 5 seconds of runtime, then there is no amount of RAM that would be enough for longer runs. And turning trace on/off (runtime/trace) wouldn't be a very nice experience if you are trying to catch a event you can't predict.

Do you know of any other similar tools but live so shouldn't eat a ton of ram?

**Comment by davecheney** [comment]
What was the program doing during that trace. Based on what I've seen the resources required to analyse a trace are proportional to the number of goroutines events during that time. Where an event is, a goroutines starting and stopping execution.

**Comment by AlexRouSg** [comment]
Running I think 4 goroutines and calling a bunch of C funcs in a loop... I guess the worst case for trace resource consumption?

**Comment by davecheney** [comment]
Pretty much.

**Comment by hyangah** [comment]
Thanks Ian for fixing the label mistake.
--
It is indeed a pathological case.

The trace contains more than 10.5million events, mostly ProcStart/ProcStop or Syscall events.
Those events are encoded with 2-3bytes each.  

When parsing the trace, each event is mapped to the trace.Event object
(https://github.com/golang/go/blob/master/src/internal/trace/parser.go#L35)
which takes hundreds of bytes. Most of the Event's fields are irrelevant to
ProcStart/ProcStop, so that implies possible optimization around the trace.Event type.

Then, the 10.5million events are mapped into 17million TraceViewer type objects  
(>150bytes each) for chrome trace viewer. This is done just to split the TraceVIewer 
events into groups, which I think can be optimized too.

Memory fragmentation makes this issue worse.

**Comment by hyangah** [comment]
According to the cl/92376 description, the memory usage went down to 3.7G (from 6.2G).
That's not as huge as we hoped but it's something. We are closing this issue here.
We will need major changes in data structure & algorithm for more memory usage reduction.
